## CIND830 - Python Programming for Data Science  
### Assignment 2 (10% of the final grade) 
### Due on March 07, 2022 11:59 PM 

*****
This is a Jupyter Notebook document that extends a simple formatting syntax for authoring HTML and PDF. Review [this](https://jupyter-notebook.readthedocs.io/en/stable/notebook.html) website for more details on using Jupyter Notebooks.

Use the JupyterHub server on the Google Cloud Platform, provided by your designated instructor, for this assignment. Ensure using **Python 3.7** release or higher then complete the assignment by inserting your Python code wherever seeing the string `#INSERT YOUR ANSWER HERE.`

When you click the `File` button, from the top navigation bar, then select `Export Notebook to HTML`, an HTML document will be generated that includes both the assignment content and the output of any embedded Python code chunks.

Use [these](https://www.ryerson.ca/courses/students/tutorials/assignments/) guidelines to submit **both** the IPYNB and the exported file (HTML). Failing to submit both files will be subject to mark deduction.

Please be advised that you cannot get more than 100% in this assignment, and the **BONUS** question (if there is any) will only be graded if all other questions have been submitted.

*****


### Question 1  **[25 pts]**:

**a)**  **[10 pts]**  Write a Python program that converts a list of email addresses to a dictionary, where the keys are the usernames, and the values are the respective domains.

For example:  If the user enters `harry@abc.com,larry@abc.ca,sally@abc.org`, then the output should be `{'harry': 'abc.com', 'larry': 'abc.ca', 'sally': 'abc.org'}`


In [ ]:
userInput1 = input("Please enter email address. For multiple email addresses, please separate with a comma: ")
emailList = userInput1.split(',')
newList = []
[newList.extend(element.split('@')) for element in emailList]

def List2Dict(emailList):
    newDict = {emailList[i]: emailList[i + 1] for i in range(0, len(emailList), 2)}
    return newDict

print(List2Dict(newList))

Please enter email address. For multiple email addresses, please separate with a comma: harry@abc.com,larry@abc.ca,sally@abc.org
{'harry': 'abc.com', 'larry': 'abc.ca', 'sally': 'abc.org'}


**b)**  **[10 pts]** Write a Python program to accept the length of a dictionary, then create key-value pairs where the key is a random integer in the range 1 to 100 and the value is the square of the integer.

For example: If the user enters `5`, then the output should be `{68: 4624, 28: 784, 99: 9801, 15: 225, 6: 36}`

In [ ]:
import random

userLength = int(input("Please enter an integer length: "))

sqDict = {}
for x in range(userLength):
    x = random.randint(1,100)
    sqDict[x] = x**2
print(sqDict)

Please enter an integer length: 5
{50: 2500, 63: 3969, 45: 2025, 99: 9801, 2: 4}


**c)**  **[5 pts]** List at least three diferences between a list and a dictionary.  Illustrate your work with simple examples.

1. A list is created by placing elements in [] separated by commas ',' while a dictionary is created by placing elements in {} as "key":"value" each key-value is a pair separated by commons ','
2. A list's elements are accessed via indices while a dictionary's elements are accessed via key-values.
2. A list's elements are mutable while dictionaries' key must be unique and immutable.
 


In [ ]:
# 1. A list is created by placing elements in [] separated by commas ',' while a 
# dictionary is created by placing elements in {} as "key":"value" each key-value 
# is a pair separated by commons ','
ListA = ['Apple', '3', 5.0, [1,2]]
DictA = {0: 'Apple', 1: '3', 'Two': 5.0, 3: [1,2]}
print("ListA: ", ListA)
print("DictA: ", DictA)

ListA:  ['Apple', '3', 5.0, [1, 2]]
DictA:  {0: 'Apple', 1: '3', 'Two': 5.0, 3: [1, 2]}


In [ ]:
# 2. A list's elements are accessed via indices while a dictionary's elements 
# are accessed via key-values.
print("ListA[2]: ", ListA[2])
print("DictA['Two']: ", DictA['Two'])

ListA[2]:  5.0
DictA['Two']:  5.0


In [ ]:
# A list's elements are mutable while dictionaries' key must be unique and immutable.
ListA[2] = 'Banana'
print(ListA)
DictB = {[1]: 'Apple', 1: '3', 'Two': 5.0, 3: [1,2]}
print(DictB)
# Returns TypeError with an unhashable type: 'list' as a key in DictB

['Apple', '3', 'Banana', 'Banana']


TypeError: ignored

*****

### Question 2  **[25 pts]**:

The data file used in this question is downloaded from (https://www.kaggle.com/aungpyaeap/supermarket-sales) and converted to a CSV file.  The [`supermarket_sales.csv`](https://app.box.com/s/f4hcfkferizntbev3ou8hso6nyfccf74) is a delimited text file containing the historical record of sales data in three supermarkets. The file contains the following columns separated by the comma character `,`
* Invoice id: Computer-generated sales slip invoice identification number
* Branch: Branch of supercenter (3 branches are available identified by A, B and C)
* City: Location of supercenters
* Customer type: Type of customers, recorded by Members for customers using member card and Normal for without member card
* Gender: Gender type of customer
* Product line: General item categorization groups - Electronic accessories, Fashion accessories, Food and beverages, Health and beauty, Home and Lifestyle, Sports and travel
* Unit price: The price of each product in $
* Quantity: Number of products purchased by the customer
* Tax: 5% tax fee for customer buying
* Total: Total price including tax
* Date: Date of purchase (Record available from January 2019 to March 2019)
* Time: Purchase time (10am to 9pm)
* Payment: Payment used by customer for purchase (3 methods are available – Cash, Credit card and Ewallet)
* COGS: Cost of goods sold
* Gross margin percentage: Gross margin percentage
* Gross income: Gross income
* Rating: Customer satisfaction rating on their overall shopping experience (On a scale of 1 to 10)

Note: Only Python [built-in functions](https://docs.python.org/3.7/library/functions.html) or [standard modules](https://docs.python.org/3.7/py-modindex.html) are allowed to be used in this question. Do not use a third-party library for any part of the code. Importing any third-party library will result in marks deduction.

**a)**  **[10 pts]**  Write a Python function that reads the Supermarket Sales data into a 2 dimensional structure.

Note: The first line of the CSV contains the column headings, therefore it should not be stored in the 2 dimensional data list. You can skip the first line of the data file

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving supermarket_sales.csv to supermarket_sales.csv


In [ ]:
import pandas as pd

def csvReader(csvFile):
  """ CSV File Reader """
  data = pd.read_csv(csvFile, skiprows = 1, header = None)
  return data

csvReader("supermarket_sales.csv")


              0  1          2       3       4                       5      6   \
0    750-67-8428  A     Yangon  Member  Female       Health and beauty  74.69   
1    226-31-3081  C  Naypyitaw  Normal  Female  Electronic accessories  15.28   
2    631-41-3108  A     Yangon  Normal    Male      Home and lifestyle  46.33   
3    123-19-1176  A     Yangon  Member    Male       Health and beauty  58.22   
4    373-73-7910  A     Yangon  Normal    Male       Sports and travel  86.31   
..           ... ..        ...     ...     ...                     ...    ...   
995  233-67-5758  C  Naypyitaw  Normal    Male       Health and beauty  40.35   
996  303-96-2227  B   Mandalay  Normal  Female      Home and lifestyle  97.38   
997  727-02-1313  A     Yangon  Member    Male      Food and beverages  31.84   
998  347-56-2442  A     Yangon  Normal    Male      Home and lifestyle  65.82   
999  849-09-3807  A     Yangon  Member  Female     Fashion accessories  88.34   

     7        8          9 

**b)**  **[5 pts]**  Write a function to receive the Supermarket Sales list (that you created in the previous question) and return the total sales (the sum of column `Total`) of the branches as a dictionary

The returned information should look like (the numbers given below may not be correct):  
{"A": 123456.123, "B": 234567.567, "C": 345678.9987}

In [ ]:
df = csvReader("supermarket_sales.csv")

def SumOfSales(data_frame):
  """ Returns the sum of values from the Total Sales column from
      the Branches as a dictionary """
  totalDf = data_frame.iloc[:, [1,9]]
  branchA = totalDf[totalDf[1] == 'A']
  branchB = totalDf[totalDf[1] == 'B']
  branchC = totalDf[totalDf[1] == 'C']
  
  salesTotal_dict = {"A": round(branchA[9].sum(), 2), "B": round(branchB[9].sum(), 2), "C": round(branchC[9].sum(), 2)}
  return salesTotal_dict

SumOfSales(df)


{'A': 106200.37, 'B': 106197.67, 'C': 110568.71}

**c)  [5 pts]** Write a function to receive the Supermarket Sales list and return an object (maybe a list or a dictionary) containing the total number of sales per Branch and per Customer Type.

The returned information can be either one of the below choices or another (The numbers given below may not be correct):)
* a dictionary like {"A": {"Member": 230, "Normal": 351}, "B": {"Member": 123, "Normal": 117},  "C": {"Member": 335, "Normal": 18}}
* a list like [ [230, 351], [123, 117], [335, 18]]

You need to consider the best structure to return since you will be displaying this information in a human-readable format in the following question. 

In [ ]:
df = csvReader("supermarket_sales.csv")

def branchSales_custType(data_frame):
  """ Returns a dictionary with the branch's total sales 
      for the customer type """
  totalDf = data_frame.iloc[:, [1,3,9]]
  bA_mem = totalDf[(totalDf[1] == 'A') & (totalDf[3] == 'Member')]
  bA_nor = totalDf[(totalDf[1] == 'A') & (totalDf[3] == 'Normal')]
  bB_mem = totalDf[(totalDf[1] == 'B') & (totalDf[3] == 'Member')]
  bB_nor = totalDf[(totalDf[1] == 'B') & (totalDf[3] == 'Normal')]
  bC_mem = totalDf[(totalDf[1] == 'C') & (totalDf[3] == 'Member')]
  bC_nor = totalDf[(totalDf[1] == 'C') & (totalDf[3] == 'Normal')]

  saleCustomer_dict = {"A": {"Member": round(bA_mem[9].sum(), 2), "Normal": round(bA_nor[9].sum(), 2)},
                       "B": {"Member": round(bB_mem[9].sum(), 2), "Normal": round(bB_nor[9].sum(), 2)},
                       "C": {"Member": round(bC_mem[9].sum(), 2), "Normal": round(bC_nor[9].sum(), 2)}}
  return saleCustomer_dict

branchSales_custType(df)

{'A': {'Member': 53637.48, 'Normal': 52562.9},
 'B': {'Member': 53704.69, 'Normal': 52492.99},
 'C': {'Member': 56881.28, 'Normal': 53687.42}}

**d) [5 pts]** Write a Python program to do the following:
* Call the function you wrote in **a)** to read the data file
* Call the function you wrote in **b)** to calculate the total sales information and display the information on the screen
* Call the function you wrote in **c)** to calculate the total number of sales per Branch and per Customer Type and display the information on the screen

Note: While displaying the information on the screen, the outputs should be human-readable and understandable (e.g. column headings, row names)

In [ ]:
def displayAll(csvFile):
  """ Displays all three predefined functions:
      1. Reads the data file
      2. Calculates the sum of the total sales
      3. Calculates the total of sales per branch per customer type """
  pt1 = csvReader(csvFile)
  pt1.columns = ["Invoice ID", "Branch", "City", "Customer", "Gender", 
                      "Product Line", "Unit Price", "Quantity", "Tax 5%", 
                      "Total", "Date", "Time", "Payment", "cogs", "Gross Margin %", 
                      "Gross Income", "Rating"]
  print("The Supermarket Sales Dataframe:")
  print(pt1)

  pt2 = csvReader(csvFile)
  pt2dict = SumOfSales(pt2)
  print("\n\nThe Total Sales for each Branch are:")
  df2 = pd.DataFrame(list(pt2dict.items()),columns = ['Branch','Total Sales $'])
  print (df2)

  print("\n\nThe Total Sales for each Branch are by Customer Type are:")
  pt3dict = branchSales_custType(pt2)
  df3 = pd.DataFrame(list(pt3dict.items()),columns = ['Branch','Customer Type & Total Sales $'])
  print(df3)


displayAll("supermarket_sales.csv")

The Supermarket Sales Dataframe:
      Invoice ID Branch       City Customer  Gender            Product Line  \
0    750-67-8428      A     Yangon   Member  Female       Health and beauty   
1    226-31-3081      C  Naypyitaw   Normal  Female  Electronic accessories   
2    631-41-3108      A     Yangon   Normal    Male      Home and lifestyle   
3    123-19-1176      A     Yangon   Member    Male       Health and beauty   
4    373-73-7910      A     Yangon   Normal    Male       Sports and travel   
..           ...    ...        ...      ...     ...                     ...   
995  233-67-5758      C  Naypyitaw   Normal    Male       Health and beauty   
996  303-96-2227      B   Mandalay   Normal  Female      Home and lifestyle   
997  727-02-1313      A     Yangon   Member    Male      Food and beverages   
998  347-56-2442      A     Yangon   Normal    Male      Home and lifestyle   
999  849-09-3807      A     Yangon   Member  Female     Fashion accessories   

     Unit Price  Q

*****

### Question 3  **[25 pts]**: 
Write a program detailing the number of passengers per vehicle.  The program includes three classes: `Vehicle`, `Car`, `Bus` where `Vehicle` is the parent class of `Car` and `Bus`. 

**a)**  **[10 pts]**  Each of the `Car` and `Bus` classes should have a constructor method `__init()__` that generates and returns a random number of passengers based on the type of the vehicle; If it is a `Bus`, then the number of passengers should be within a range of 5 and 20.  If the vehicle is a `Car`, the range should be between 1 to 4 passengers, inclusively.

**b)**  **[10 pts]**  The program's `main()` function would display the following output.  Notably, since the numbers of passengers are randomly selected, the program will keep changing them every time it runs.

```
--------------------------------
Bus Passengers:   64
Car Passengers:   13
Total Passengers: 77
--------------------------------
Bus#:  1  Number of Passengers :18
Bus#:  2  Number of Passengers :20
Bus#:  3  Number of Passengers : 6
Bus#:  4  Number of Passengers : 9
Bus#:  5  Number of Passengers :11
Car#:  6  Number of Passengers : 1
Car#:  7  Number of Passengers : 2
Car#:  8  Number of Passengers : 4
Car#:  9  Number of Passengers : 2
Car#: 10  Number of Passengers : 4
--------------------------------
```

**c)**  **[5 pts]** Restructure the program by re-defining the `Bus` class to be the superclass of `Car`, and display the same output.

In [2]:
# a) 
import random

class Vehicle():
  def __init__(self, type):
    self.type = type

  def submethod(self, Bus):
    print(self.type)

class Car(Vehicle):
  def __init__(self, number, capacity=None):
    if capacity is None:
      capacity = random.randint(1,4)
    self.number = number
    self.capacity = capacity

class Bus(Vehicle):
  def __init__(self, number, capacity=None):
    if capacity is None:
      capacity = random.randint(5,20)    
    self.number = number
    self.capacity = capacity

In [3]:
# b) 
def main():
  Bus1 = Bus(1)
  Bus2 = Bus(2)
  Bus3 = Bus(3) 
  Bus4 = Bus(4)
  Bus5 = Bus(5)
  Car6 = Car(6)
  Car7 = Car(7)
  Car8 = Car(8) 
  Car9 = Car(9)
  Car10 = Car(10)
  
  Busses = [Bus1.capacity, Bus2.capacity, Bus3.capacity, Bus4.capacity, Bus5.capacity]
  Cars = [Car6.capacity, Car7.capacity, Car8.capacity, Car9.capacity, Car10.capacity]
  sumBus = sum(Busses)
  sumCar = sum(Cars)
    
  print('---------------------------------------------')
  print('Bus Passengers:\t\t\t\t', sumBus)
  print('Car Passengers:\t\t\t\t', sumCar)
  print('Total Passengers:\t\t\t', (sumBus + sumCar))
  print('---------------------------------------------')
 
  countB = 1
  for element in Busses:
    print("Bus#:\t", countB, "\tNumber of Passengers:\t", element)
    countB += 1

  countC = 6
  for element in Cars:
    print("Car#:\t", countC, "\tNumber of Passengers:\t", element)
    countC += 1
  print('---------------------------------------------')

if __name__ == "__main__":
  main()

---------------------------------------------
Bus Passengers:				 58
Car Passengers:				 14
Total Passengers:			 72
---------------------------------------------
Bus#:	 1 	Number of Passengers:	 20
Bus#:	 2 	Number of Passengers:	 7
Bus#:	 3 	Number of Passengers:	 8
Bus#:	 4 	Number of Passengers:	 12
Bus#:	 5 	Number of Passengers:	 11
Car#:	 6 	Number of Passengers:	 4
Car#:	 7 	Number of Passengers:	 1
Car#:	 8 	Number of Passengers:	 4
Car#:	 9 	Number of Passengers:	 2
Car#:	 10 	Number of Passengers:	 3
---------------------------------------------


In [ ]:
import random

class Vehicle():
  def __init__(self, type):
    self.type = type

  def submethod(self, Bus):
    print(self.type)

class Car(Vehicle):
  def __init__(self, number, capacity=None):
    if capacity is None:
      capacity = random.randint(1,4)
    self.number = number
    self.capacity = capacity

class Bus(Vehicle):
  def __init__(self, number, capacity=None):
    super().__init__(number)
    if capacity is None:
      capacity = random.randint(5,20)    
    self.number = number
    self.capacity = capacity

def main():
  Bus1 = Bus(1)
  Bus2 = Bus(2)
  Bus3 = Bus(3) 
  Bus4 = Bus(4)
  Bus5 = Bus(5)
  Car6 = Car(6)
  Car7 = Car(7)
  Car8 = Car(8) 
  Car9 = Car(9)
  Car10 = Car(10)
  
  Busses = [Bus1.capacity, Bus2.capacity, Bus3.capacity, Bus4.capacity, Bus5.capacity]
  Cars = [Car6.capacity, Car7.capacity, Car8.capacity, Car9.capacity, Car10.capacity]
  sumBus = sum(Busses)
  sumCar = sum(Cars)
    
  print('---------------------------------------------')
  print('Bus Passengers:\t\t\t\t', sumBus)
  print('Car Passengers:\t\t\t\t', sumCar)
  print('Total Passengers:\t\t\t', (sumBus + sumCar))
  print('---------------------------------------------')
 
  countB = 1
  for element in Busses:
    print("Bus#:\t", countB, "\tNumber of Passengers:\t", element)
    countB += 1

  countC = 6
  for element in Cars:
    print("Car#:\t", countC, "\tNumber of Passengers:\t", element)
    countC += 1
  print('---------------------------------------------')

if __name__ == "__main__":
  main()

---------------------------------------------
Bus Passengers:				 77
Car Passengers:				 16
Total Passengers:			 93
---------------------------------------------
Bus#:	 1 	Number of Passengers:	 8
Bus#:	 2 	Number of Passengers:	 18
Bus#:	 3 	Number of Passengers:	 19
Bus#:	 4 	Number of Passengers:	 16
Bus#:	 5 	Number of Passengers:	 16
Car#:	 6 	Number of Passengers:	 3
Car#:	 7 	Number of Passengers:	 4
Car#:	 8 	Number of Passengers:	 4
Car#:	 9 	Number of Passengers:	 2
Car#:	 10 	Number of Passengers:	 3
---------------------------------------------


### Question 4  **[25 pts]**:

**a)**  **[5 pts]** Define and implement a function that accepts a tuple of lists and returns the sum of only the unique numbers in that tuple.

For example, if the user enters the following tuple: `
([1, 5], [7, 8, 3, 2], [3, 3, 2])`
then the output should be `26`.

In [ ]:
import re

def sumUniqueTup():
  tupleList = input("Please enter a tuple of lists: ")
  temp = re.sub(r'[\[\]\(\), ]', '', str(tupleList))
  uniqueVals = [int(ele) for ele in set(temp)]
  print("The sum of unique numbers in the tuple are: " + str(sum(res)))

sumUniqueTup()

**b)**  **[5 pts]** Define and implement a function that converts the a list of tuples to a dictionary where the first value of each tuple is the key, and the second value is the dictionary value. <br>

For example, if the user enters the following list: 
`
[('Person1', 80), ('Person2', ), ('Person3', 20), ('Person4', 20, 60)]
`

Then the output should be: `{'Person1': 80, 'Person2': None, 'Person3': 20, 'Person4': 20}`


In [6]:
def Convert(tup, di):
    tups = input("Enter a list of tuples: ")
    dictionary = {}
    for a, b in tups:
        di.setdefault(a, []).append(b)
    print(dictionary)




**c)**  **[15 pts]** The following script extracts the text from the `https://norvig.com/big.txt` file, stores them in a string object, then assigns this string to a variable called `data` using the [requests](https://pypi.org/project/requests/) third-party library.
```
#!pip install requests
import requests
response = requests.get('https://norvig.com/big.txt')
data = response.text
```
Run the script above, uncomment the first line if you need to install the library on your machine, then define and implement a function that returns the followings:

* The total number of unique words and unique characters
* The top ten most frequent words and the number of times these words appear in the text.

In [ ]:
!pip install requests
import requests
response = requests.get('https://norvig.com/big.txt')
data = response.text

In [ ]:
from collections import Counter

def count_unqiue_words(data_file):
  countWords = 0
  u_words = set(data_file.split())
  for word in u_words:
    countWords += 1
  print('Total Unique Words:', countWords)

def count_unique_characters(data_file):
  u_char = set(data_file)
  print('Total Unique Characters:', len(u_char))

def topTenWords(data_file):
  u_words = data_file.split()
  wCounter = Counter(u_words)
  topTen = wCounter.most_common(10)
  print("Word\t", "Frequency")
  for i,j in topTen:
    print(i,"\t",j)

def main():
  count_unqiue_words(data)
  count_unique_characters(data)
  topTenWords(data)

if __name__ == "__main__":
  main()
  

Total Unique Words: 81397
Total Unique Characters: 93
Word	 Frequency
the 	 71744
of 	 39169
and 	 35968
to 	 27895
a 	 19811
in 	 19515
that 	 11216
was 	 11129
his 	 9561
he 	 9362


#### This is the end of assignment 2